modules/service/src/main/scala/lucuma/odb/service/ProposalAttachmentMetadataService.scala

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2016-2023 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package lucuma.odb.service
+
+import cats.data.NonEmptyList
+import cats.effect.Sync
+import cats.syntax.all.*
+import edu.gemini.grackle.Result
+import lucuma.core.model.ObsAttachment
+import lucuma.core.model.Program
+import lucuma.core.model.User
+import lucuma.odb.data.Nullable
+import lucuma.odb.data.Tag
+import lucuma.odb.graphql.input.ProposalAttachmentPropertiesInput
+import lucuma.odb.util.Codecs.*
+import natchez.Trace
+import skunk.*
+import skunk.codec.all.*
+import skunk.implicits.*
+
+trait ProposalAttachmentMetadataService[F[_]] {
+  def updateProposalAttachments(
+    programId: Program.Id,
+    SET: ProposalAttachmentPropertiesInput.Edit,
+    which: AppliedFragment
+  ): F[List[Tag]]
+}
+
+object ProposalAttachmentMetadataService {
+
+  def fromSessionAndUser[F[_]: Sync: Trace](
+    session: Session[F],
+    user:    User
+  ): ProposalAttachmentMetadataService[F] =
+    new ProposalAttachmentMetadataService[F] {
+      def updateProposalAttachments(
+          programId: Program.Id,
+          SET: ProposalAttachmentPropertiesInput.Edit, 
+          which: AppliedFragment
+      ): F[List[Tag]] = 
+        Statements.updateProposalAttachments(programId, SET, which).fold(Nil.pure[F]) { af =>
+          session.prepareR(af.fragment.query(tag)).use { pq =>
+            pq.stream(af.argument, chunkSize = 1024).compile.toList
+          }
+        }
+    }
+
+  object Statements {
+
+    def updates(SET: ProposalAttachmentPropertiesInput.Edit): Option[NonEmptyList[AppliedFragment]] = {
+      val upDescription = sql"c_description = ${text_nonempty.opt}"
+      val upChecked     = sql"c_checked = $bool"
+      
+      NonEmptyList.fromList(
+        List(
+          SET.description match {
+            case Nullable.Null           => Some(upDescription(None))
+            case Nullable.Absent         => None
+            case Nullable.NonNull(value) => Some(upDescription(Some(value)))
+          },
+          SET.checked.map(upChecked)
+        ).flatten
+      )
+    }
+
+    def updateProposalAttachments(
+      programId: Program.Id,
+      SET: ProposalAttachmentPropertiesInput.Edit,
+      which: AppliedFragment
+    ): Option[AppliedFragment] = 
+      updates(SET).map {us =>
+        void"UPDATE t_proposal_attachment "  |+|
+        void"SET " |+| us.intercalate(void", ") |+| void" " |+|
+        sql"WHERE t_proposal_attachment.c_program_id = $program_id ".apply(programId) |+|
+        void"AND t_proposal_attachment.c_attachment_type IN (" |+| which |+| void") " |+|
+        // void"AND (t_proposal_attachment.c_attachment_type, t_proposal_attachment.c_program_id) IN (" |+| which |+| void") " |+|
```

Comment:
If I mark `program_id` as a key in the `ProposalAttachmentMapping`, I would need to use this IN clause with both the attachment_type and program_id. I'm not sure why the order is attachment_type first.

Reply:
The `which` AppliedFragment is created in `MutationMapping.scala` starting at line 150. Grackle somehow uses the filterPredicate to build the AppliedFragment using the keys defined in the table mapping.